finetune/utils.py

```diff
@@ -407,3 +407,58 @@ def indico_to_finetune_sequence(texts, labels=None, multi_label=True, none_value
         all_idxs.append(doc_current_label_idx)
 
     return all_subseqs, all_labels, all_association_type, all_association_idx, all_idxs
+
+
+def get_grad_accumulation_optimizer(optimizer_class, accum_steps):
+    """
+    Adds gradient accumulation to an Optimizer.
+
+    :param optimizer_class: A subclass of tf.train.Optimizer to add gradient accumulation too.
+    :param accum_steps: An int value determining how many gradients to accumulate before performing an optimizer step.
+
+    :return: A new Optimizer, with gradient accumulation.
+    """
+
+    class GradAccumulationOptimiser(optimizer_class):
+        def __init__(self, *args, **kwargs):
+            super().__init__(*args, **kwargs)
+
+        def apply_gradients(self, grads_and_vars, global_step=None, *args, **kwargs):
+            add_gradients_ops = []
+            zero_ops = []
+            grads_and_accumulated_vars = []
+            for g, v in grads_and_vars:
+                accum_grad = tf.get_variable(
+                    name=g.name[:-2] + "_acc",
+                    shape=g.shape,
+                    dtype=g.dtype,
+                    initializer=tf.constant_initializer(0)
+                )
+
+                add_gradients_ops.append(accum_grad.assign_add(g))
+                zero_ops.append(accum_grad.initializer)
+
+                grads_and_accumulated_vars.append((accum_grad, v))
+
+            global_step = global_step or tf.train.get_or_create_global_step()
+
+            def apply_grads():
+                with tf.control_dependencies([
+                    super(GradAccumulationOptimiser, self).apply_gradients(grads_and_accumulated_vars, global_step=global_step, *args, **kwargs)
+                ]):
+                    apply_grads_op = tf.group(zero_ops)
```

Comment:
Also, I believe I understand why we need tf.control_dependencies(add_gradients_ops) down below--i.e. we want to make sure the gradients all get added before applying them if we do meet the condition. Why do we want tf.control_dependencies before grouping the zero_ops together? Also, why do we need to group zero_ops?

Reply:
This is a slight misuse of control_dependancies. But basically we want to apply gradients and then and only then can we reset the accumulation variables and I want a single operation that performs that sequence of events.